### PR TITLE
feat: Support runtime metrics

### PR DESCRIPTION
**Pull request by adebayor123**
*Dependent PR*:
- https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/243

*Relevant PR*
- https://github.com/aws/amazon-cloudwatch-agent-operator/pull/229
- https://github.com/aws-observability/helm-charts/pull/97

# Description of the issue
ApplicationSignals is expanding the set of monitored metrics to include runtime metrics. 

# Description of changes
This pull request contains changes required to support runtime metrics in CWAgent for ApplicationSignals. To enable the change, we are introducing a new processor, `metrictransformprocessor`, and expanding the exporter config to handle JVM and Python runtime metrics.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
`make test` & manual testing

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

File: translator/translate/otel/exporter/awsemf/appsignals_config_generic.yaml [hunk 1]
```diff
@@ -30,4 +30,12 @@ metric_declarations:
...
       - Fault
-      - Error
\ No newline at end of file
+      - Error
+  - dimensions:
+      - [Environment, Service]
+    label_matchers:
+      - label_names:
+          - Telemetry.Source
+        regex: '^RuntimeMetrics$'
```
Comment (bjrara) [inline translator/translate/otel/exporter/awsemf/appsignals_config_generic.yaml:39]: RuntimeMetric

**Review by bjrara — COMMENTED** [review]
Please evaluate whether to use `mean` or `sum` in the `aggregation_type`.

File: translator/translate/otel/processor/appsignalsmetricstransformprocessor/config/appSignalsRuntimeConfig.yaml [hunk 2]
```diff
@@ -0,0 +1,243 @@
+transforms:
+  - include: jvm.cpu.recent_utilization
+    action: insert
+    new_name: JVMCpuRecentUtilization
+    operations:
+      - action: aggregate_labels
```
[the remark this answers is not in the document]
Comment (adebayor123) [reply translator/translate/otel/processor/appsignalsmetricstransformprocessor/config/appSignalsRuntimeConfig.yaml:6]: No, we didn't receive multiple but I am using aggregate to ensure all metric labels would be gone prior to adding Telemetry.Source

File: plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer.go [hunk 3]
```diff
@@ -121,6 +121,10 @@ func (n *attributesNormalizer) copyResourceAttributesToAttributes(attributes, re
 			}
 		}
 	}
+	// runtime metrics do not have service attribute, so need to manually add
+	if serviceAttribute, ok := resourceAttributes.Get("aws.local.service"); ok {
+		attributes.PutStr("aws.local.service", serviceAttribute.AsString())
```
[the remark this answers is not in the document]
Comment (adebayor123) [reply plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer.go:126]: Done!

File: plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer.go [hunk 4]
```diff
@@ -121,6 +121,10 @@ func (n *attributesNormalizer) copyResourceAttributesToAttributes(attributes, re
 			}
 		}
 	}
+	// runtime metrics do not have service attribute, so need to manually add
+	if serviceAttribute, ok := resourceAttributes.Get("aws.local.service"); ok {
```
Comment (bjrara) [inline plugins/processors/awsapplicationsignals/internal/normalizer/attributesnormalizer.go:125]: Please use [`attributes.AWSLocalService`](https://github.com/aws/amazon-cloudwatch-agent/blob/1e0a0495a68b9f808650702c20db5c80cee014a7/plugins/processors/awsapplicationsignals/internal/attributes/attributes.go#L9) instead.

File: plugins/processors/awsapplicationsignals/internal/prune/metric_pruner.go [hunk 5]
```diff
@@ -21,6 +21,9 @@ func (p *MetricPruner) ShouldBeDropped(attributes pcommon.Map) (bool, error) {
 				return true, errors.New("Metric attribute " + attributeKey + " must contain only ASCII characters.")
 			}
 		}
+		if _, ok := attributes.Get(common.MetricAttributeTelemetrySource); !ok {
+			return true, errors.New("metric must contain Telemetry.Source")
```
Comment (bjrara) [inline plugins/processors/awsapplicationsignals/internal/prune/metric_pruner.go:25]: Please capitalize the first character for the error message.

File: translator/translate/otel/processor/appsignalsmetricstransformprocessor/translator.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+	"github.com/aws/amazon-cloudwatch-agent/translator/translate/otel/common"
+)
+
+//go:embed config/appSignalsRuntimeConfig.yaml
+var appSignalsRuntimeConfig string
+
+type translator struct {
+	name     string
+	dataType component.DataType
```
Comment (sky333999) [inline translator/translate/otel/processor/appsignalsmetricstransformprocessor/translator.go:22]: Do you need this? Dont see this being used anywhere.

File: translator/translate/otel/processor/appsignalsmetricstransformprocessor/config/appSignalsRuntimeConfig.yaml [hunk 7]
```diff
@@ -0,0 +1,279 @@
+transforms:
+  - include: jvm.cpu.recent_utilization
+    action: update
+    new_name: JVMCpuRecentUtilization
+    operations:
+      - action: aggregate_labels
+        label_set: []
+        aggregation_type: sum
+      - action: add_label
+        new_label: Telemetry.Source
```
Comment (sky333999) [inline translator/translate/otel/processor/appsignalsmetricstransformprocessor/config/appSignalsRuntimeConfig.yaml:10]: Wouldnt it be cleaner to set these on the java agent when exporting them?

Comment (bjrara) [reply translator/translate/otel/processor/appsignalsmetricstransformprocessor/config/appSignalsRuntimeConfig.yaml:10]: The attribute is added so the metrics can be selected by the EMF exporter to generate metrics.
Ref:
https://github.com/aws/amazon-cloudwatch-agent/blob/bde3bd9775ae1d4e4f8a2fdb92d7b6fdd5186fba/translator/tocwconfig/sampleConfig/base_appsignals_config.yaml#L23-L25

File: translator/translate/otel/processor/appsignalsmetricstransformprocessor/translator.go (hunk 6 above)
Comment (bjrara) [reply translator/translate/otel/processor/appsignalsmetricstransformprocessor/translator.go]: Removed.

File: translator/translate/otel/pipeline/applicationsignals/translator_test.go [hunk 8]
```diff
@@ -145,7 +145,7 @@ func TestTranslatorMetricsForKubernetes(t *testing.T) {
 			},
 			want: &want{
 				receivers:  []string{"otlp/application_signals"},
-				processors: []string{"resourcedetection", "awsapplicationsignals"},
+				processors: []string{"metricstransform", "resourcedetection", "awsapplicationsignals"},
```
[the remark this answers is not in the document]
Comment (bjrara) [reply translator/translate/otel/pipeline/applicationsignals/translator_test.go]: Done.

File: translator/translate/otel/processor/appsignalsmetricstransformprocessor/translator.go
[the remark this answers is not in the document]
Comment (bjrara) [reply translator/translate/otel/processor/appsignalsmetricstransformprocessor/translator.go:1]: Done.

File: translator/tocwconfig/sampleConfig/appsignals_and_ecs_config.yaml
Comment (jefchien) [inline translator/tocwconfig/sampleConfig/appsignals_and_ecs_config.yaml:1]: Looks like you might need to regenerate these.

File: translator/translate/otel/pipeline/applicationsignals/translator_test.go [hunk 9]
```diff
@@ -125,7 +125,7 @@ func TestTranslatorMetricsForKubernetes(t *testing.T) {
 			},
 			want: &want{
 				receivers:  []string{"otlp/application_signals"},
-				processors: []string{"resourcedetection", "awsapplicationsignals"},
+				processors: []string{"metricstransform/application_signals", "resourcedetection", "awsapplicationsignals"},
```
Comment (jefchien) [inline translator/translate/otel/pipeline/applicationsignals/translator_test.go:128]: nit: I know it's unrelated, but could we add `application_signals` to the `resourcedetection` processor as well? Could be done in a separate PR if you'd like.

File: translator/translate/otel/processor/metricstransformprocessor/appsignals_runtime_config.yaml
Comment (jefchien) [inline translator/translate/otel/processor/metricstransformprocessor/appsignals_runtime_config.yaml:1]: Have we tested this E2E to make sure these all result in the metrics we want?

File: translator/tocwconfig/sampleConfig/appsignals_and_ecs_config.yaml
Comment (bjrara) [reply translator/tocwconfig/sampleConfig/appsignals_and_ecs_config.yaml:1]: Yes. Regenerated.

File: translator/translate/otel/pipeline/applicationsignals/translator_test.go (hunk 9 above)
Comment (bjrara) [reply translator/translate/otel/pipeline/applicationsignals/translator_test.go:128]: I will do in a follow-up PR.

File: translator/translate/otel/processor/metricstransformprocessor/appsignals_runtime_config.yaml
Comment (bjrara) [reply translator/translate/otel/processor/metricstransformprocessor/appsignals_runtime_config.yaml:1]: Yes. https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/11492908177

**Review by sky333999 — DISMISSED** [review]

**Review by jefchien — APPROVED** [review]

**Comment by bjrara** [comment]
E2E validation after rebasing:
* Java: https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/11585089516
* Python: https://github.com/aws-observability/aws-application-signals-test-framework/actions/runs/11585085944

**Review by musa-asad — APPROVED** [review]

**Review by sky333999 — APPROVED** [review]

**Pull request merged**